### ISSUE TITLE
Make it possible to determine which guard failed on a transition

### ISSUE THREAD

**Issue by timrogers** [gocardless/statesman#254]
## Problem

Using `Machine.guard_transition` in your state machine, you can create a "guard" which determines whether a specified transition is permitted, specifying a block to be called which returns a boolean representing whether the request transition is allowed. 

But if you have multiple guards and a transition is prevented, you can't tell which guard (and thus what logic) blocked the transition.

For example, I might have a piece of code like the following, specifying a few different guards determining whether I can transition a subscription to an "active" state:

```ruby
guard_transition(from: :cancelled, to: :active) do |subscription|
  subscription.product.reactivation_permitted?
end

guard_transition(to: :active) do |subscription|
  subscription.user.payment_method.present?
end
```

Let's say for example, that I provide an API endpoint for activating a subscription. In that endpoint's code, I want to return a helpful error explaining why I couldn't transition it to `:active` if the transition fails.

At present, you can't really do this in a sensible way. Either you'll call `Machine#transition_to` which will return `false`, or `Machine#transition_to!` which will raise a `GuardFailedError`. 

At the moment, probably the best solution to this would be to duplicate the guard logic elsewhere, having some code like this:

```ruby
return reactivation_not_permitted_error unless subscription.product.reactivation_permitted?
return no_payment_method_error unless subscription.user.payment_method.present?

subscription.transition_to!(:active)
render @subscription
```

Potentially, the logic which you end up duplicating could be fairly complex, involving separate classes or multiple chained conditions.

## Potential solutions

It'd be really nice if Statesman offered a native way to solve this problem, providing some kind of context as to what failed, rather than just a blanket, generic failure.

Here are two proposals for doing this for discussion - feedback and alternatives are welcomed:

### 1. Giving guards a name

Guards could be defined with a name:

```ruby
guard_transition(to: :active, name: :payment_method_required) do |subscription|
  subscription.user.payment_method.present?
end

begin
  subscription.transition_to!(:active)
rescue Statesman::GuardFailedError => e
  return reactivation_not_permitted_error if e.name == :reactivation_must_be_permitted
  return no_payment_method_error if e.name == :payment_method_required
end

render @subscription
```

On failure, `#transition_to!` raise a `GuardFailedError` with the `:name` attached as metadata. `#transition_to` could also potentially return this too, perhaps by returning the symbol name.

### Providing a exception to guards to be raised

When defining a guard, a subclass of `GuardFailedError` could be specified to be raised if the guard fails:

```ruby
class ReactivationNotPermittedError < Statesman::GuardFailed; end
class PaymentMethodRequiredError < Statesman::GuardFailed; end

guard_transition(from: :cancelled, to: :active, error_class: ReactivationNotPermittedError) do |subscription|
  subscription.product.reactivation_permitted?
end

guard_transition(to: :active, error_class: PaymentMethodRequiredError) do |subscription|
  subscription.user.payment_method.present?
end

begin
  subscription.transition_to!(:active)
rescue ReactivationNotPermittedError
  return reactivation_not_permitted_error
rescue PaymentMethodRequiredError
  return no_payment_method_error
end

render @subscription
```

The clearest issue with this approach is that it wouldn't evidently be compatible with `transition_to` with its exception-free semantics.

**Comment by timrogers** [comment]
@Sinjo @hmac Interested in your thoughts! (And those of anyone else ❤️)

**Comment by hmac** [comment]
I think this is pretty valid, although I can imagine a counterargument:

If you have complex logic guarding a transition, you can extract that out into a "validator" of sorts
```ruby
class SubscriptionReactivation < Validator
  def permitted?(subscription)
    if # complex logic
      true
    else
      false
    end
  end
end

# In the API endpoint:

unless SubscriptionReactivation.permitted?(subscription)
  return reactivation_not_permitted_error
end

guard_transition(from: :cancelled, to: :active) do
  SubscriptionReactivation.permitted?(subscription)
end
```

This avoids the duplication of logic, but it's still a bit annoying that you have to run the validator twice, and this might be a real problem if it's computationally expensive or executes database queries.

Given the two options, I'd be in favour of the first. Exceptions raised from existing (unnamed) guards without names would just have `e.name == nil`, so I think it would be backwards compatible.

I'm uncomfortable with changing the return value of `#transition_to`, since anything other than `false` might be interpreted as `true` by some callers (`!!:sym == true`). We could maybe add a `#last_transition_error` method which returns the exception that would have been raised by the last call to `#transition_to`, but I'm not a huge fan of that either because it introduces some state that could cause nasty bugs.

**Comment by timrogers** [comment]
@hmac Shall I go ahead and implement no. 1 at some point? Feels like it'd be quite nice, and it's not a breaking change.

**Comment by JoeSouthan** [comment]
Fixed in #499
